File: Pastura/PasturaTests/LLM/PartialOutputExtractorTests.swift
import Foundation
import Testing

@testable import Pastura

/// Exercises the state-machine partial-JSON extractor against the
/// synthetic byte-stream fixtures from item 1 and additional edge cases.
/// Every test feeds the buffer incrementally (byte-by-byte where it
/// matters) and asserts the extracted snapshot at each prefix is
/// consistent with what the canonical `JSONResponseParser` would return
/// on the final buffer.
@Suite(.timeLimit(.minutes(1)))
struct PartialOutputExtractorTests {

  let extractor = PartialOutputExtractor()

  // MARK: - Empty / pre-JSON

  @Test func emptyBufferYieldsEmptySnapshot() {
    #expect(extractor.extract(from: "") == .empty)
  }

  @Test func bufferWithoutOpenBraceYieldsEmpty() {
    #expect(extractor.extract(from: "Sure! Here is the JSON:") == .empty)
  }

  @Test func bufferStoppedBeforeOpeningQuoteYieldsEmpty() {
    // Colon not yet arrived — extractor must wait.
    #expect(extractor.extract(from: #"{"statement""#).primary == nil)
    // Colon arrived but no opening quote — still wait.
    #expect(extractor.extract(from: #"{"statement":"#).primary == nil)
  }

  // MARK: - Primary key reveals

  @Test func primaryRevealsAfterOpeningQuote() {
    // Opening quote present but no content — primary is empty string, not nil.
    let snap = extractor.extract(from: #"{"statement":""#)
    #expect(snap.primary == "")
  }

  @Test func primaryRevealsIncrementalText() {
    let snap = extractor.extract(from: #"{"statement":"Let's coope"#)
    #expect(snap.primary == "Let's coope")
    #expect(snap.thought == nil)
  }

  @Test func primaryRevealsCompleteValue() {
    let snap = extractor.extract(
      from: #"{"statement":"Let's cooperate."}"#)
    #expect(snap.primary == "Let's cooperate.")
  }

  @Test func allKnownPrimaryKeysResolve() {
    for key in PartialOutputExtractor.primaryKeys {
      let snap = extractor.extract(from: #"{"\#(key)":"val"}"#)
      #expect(snap.primary == "val", "primary key \(key) failed to resolve")
    }
  }

  // MARK: - Thought

  @Test func thoughtResolvesAfterPrimary() {
    let snap = extractor.extract(
      from: #"{"statement":"hi","inner_thought":"secret"}"#)
    #expect(snap.primary == "hi")
    #expect(snap.thought == "secret")
  }

  @Test func thoughtIsNilIfNotYetOpened() {
    let snap = extractor.extract(from: #"{"statement":"hi""#)
    #expect(snap.primary == "hi")
    #expect(snap.thought == nil)
  }

  // MARK: - Escape handling

  @Test func escapedQuoteInPrimary() {
    let snap = extractor.extract(
      from: #"{"statement":"She said \"hi\""}"#)
    #expect(snap.primary == #"She said "hi""#)
  }

  @Test func escapedBackslashInPrimary() {
    let snap = extractor.extract(
      from: #"{"statement":"a\\b"}"#)
    #expect(snap.primary == #"a\b"#)
  }

  @Test func escapedNewlineInPrimary() {
    let snap = extractor.extract(
      from: #"{"statement":"line1\nline2"}"#)
    #expect(snap.primary == "line1\nline2")
  }

  @Test func incompleteEscapeAtEndHoldsBack() {
    // Buffer ends with a lone backslash — the next char might be `"` (end
    // of string) or `\\` (literal backslash). Must not emit the `\` yet.
    let snap = extractor.extract(from: #"{"statement":"x\"#)
    #expect(snap.primary == "x")
  }

  @Test func incompleteUnicodeEscapeHoldsBack() {
    // \uXXXX needs 4 hex digits — fewer is incomplete.
    let snap = extractor.extract(from: #"{"statement":"a\u00"#)
    #expect(snap.primary == "a")
  }

  @Test func completeUnicodeEscapeDecodes() {
    // \u00E9 is é.
    let snap = extractor.extract(
      from: #"{"statement":"caf\u00e9"}"#)
    #expect(snap.primary == "café")
  }

  // MARK: - Thinking-tag handling

  @Test func unclosedChannelThinkingTagHidesEverything() {
    // While we're inside a thinking tag, no extraction should fire even
    // if `{` appears later in the tag's content.
    let snap = extractor.extract(
      from: #"<|channel>thought\nI'll say {"statement":"hi"}"#)
    #expect(snap == .empty)
  }

  @Test func closedChannelThinkingTagIsStripped() {
    let buffer = """
      <|channel>thought
      reasoning here
      <channel|>{"statement":"visible"}
      """
    #expect(extractor.extract(from: buffer).primary == "visible")
  }

  @Test func unclosedThinkTagHidesEverything() {
    let snap = extractor.extract(
      from: #"<think>reasoning {"statement":"hi"#)
    #expect(snap == .empty)
  }

  // MARK: - Leading garbage

  @Test func leadingGarbageSkippedUntilBrace() {
    let snap = extractor.extract(
      from: "Sure! Here's my response:\n\n{\"statement\":\"OK\"}")
    #expect(snap.primary == "OK")
  }

  // MARK: - Byte-stream replay (item 1 fixtures)
  //
  // Feeding every prefix of the reconstructed bytes must never produce a
  // snapshot that contradicts what the final (complete) parse would
  // reveal. "Contradict" means: primary text at prefix N is not a
  // prefix of primary at prefix N+1, or the final primary.

  @Test func cjkBoundaryFixtureReplaysWithoutContradiction() throws {
    let fixture = try LlamaCppTraceFixtures.named("cjk_utf8_boundary")
    try replayAndCheckMonotonicity(fixture: fixture)
  }

  @Test func escapedQuoteFixtureReplaysWithoutContradiction() throws {
    let fixture = try LlamaCppTraceFixtures.named("escaped_quote")
    try replayAndCheckMonotonicity(fixture: fixture)
  }

  @Test func normalFixtureReplaysWithoutContradiction() throws {
    let fixture = try LlamaCppTraceFixtures.named("normal_statement")
    try replayAndCheckMonotonicity(fixture: fixture)
  }

  // MARK: - Replay helper

  /// Feeds `fixture` byte-by-byte through the extractor and verifies:
  /// 1. Primary text never shrinks between successive prefixes.
  /// 2. Primary text at every prefix is a prefix of the canonical final
  ///    parse result (when the final parse succeeds).
  /// 3. No replacement characters (`U+FFFD`) appear in the primary at any
  ///    intermediate state — feeding a partial UTF-8 sequence must not
  ///    surface mojibake.
  private func replayAndCheckMonotonicity(
    fixture: LlamaCppTraceFixture
  ) throws {
    var bytes = Data()
    var lastPrimary = ""

    // Canonical parse for the complete text — the extractor must stay
    // consistent with this.
    let parser = JSONResponseParser()
    let finalParsed = try? parser.parse(fixture.finalText)
    let canonicalPrimary =
      finalParsed?.statement
      ?? finalParsed?.action
      ?? finalParsed?.vote

    for piece in fixture.pieces {
      for byte in piece.bytes {
        bytes.append(byte)
        // Mirror `LlamaCppService.longestValidUtf8Prefix`: feed the
        // extractor the longest valid UTF-8 prefix, not the raw bytes
        // converted with fallback-to-empty. Otherwise hitting a
        // continuation byte would briefly zero the buffer and the
        // monotonicity check would fire spuriously.
        let text = Self.longestValidUtf8Prefix(bytes)
        let snap = extractor.extract(from: text)
        let current = snap.primary ?? ""

        #expect(
          !current.contains("\u{FFFD}"),
          "Replacement character leaked into extracted primary at byte offset \(bytes.count)"
        )
        #expect(
          current.hasPrefix(lastPrimary) || lastPrimary.isEmpty,
          "Primary shrank: was \(lastPrimary.debugDescription), now \(current.debugDescription)"
        )
        if let canonicalPrimary {
          #expect(
            canonicalPrimary.hasPrefix(current),
            "Intermediate primary \(current.debugDescription) is not a prefix of canonical \(canonicalPrimary.debugDescription)"
          )
        }
        lastPrimary = current
      }
    }
  }

  /// Longest UTF-8-decodable prefix of `bytes`. Duplicates (intentionally
  /// — this test helper stays test-local) the production helper in
  /// `LlamaCppService`.
  private static func longestValidUtf8Prefix(_ bytes: Data) -> String {
    for trim in 0...min(3, bytes.count) {
      let slice = bytes.prefix(bytes.count - trim)
      if let text = String(data: slice, encoding: .utf8) {
        return text
      }
    }
    return ""
  }
}

File: shared/engine/src/commonTest/kotlin/com/pastura/engine/PartialOutputExtractorTests.kt
package com.pastura.engine

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

/**
 * Kotlin siblings of the Swift `PartialOutputExtractorTests` deterministic
 * cases. Feeds progressively-longer buffers through the extractor and asserts
 * the extracted `(primary, thought)` snapshot at each prefix.
 *
 * **Knowingly absent** (parity carve-out, cf. the EngineLogger behavioral
 * carve-out): the three byte-stream fixture-replay tests
 * (`cjkBoundaryFixtureReplaysWithoutContradiction`,
 * `escapedQuoteFixtureReplaysWithoutContradiction`,
 * `normalFixtureReplaysWithoutContradiction`) and their
 * `replayAndCheckMonotonicity` / `longestValidUtf8Prefix` helpers. They depend
 * on Swift `LlamaCppTraceFixtures` + the canonical `JSONResponseParser` replay
 * path, neither of which is in the Kotlin parity scope.
 *
 * ### Kotlin string-literal note
 *
 * Buffer literals use triple-quoted raw strings (`"""..."""`) — like the Swift
 * original's `#"..."#` raw strings, they do NOT process escapes, so a `\n` /
 * `é` / `\\` in the buffer reaches the extractor as LITERAL backslash-escape
 * bytes for it to decode. Regular double-quoted strings are used only for
 * assertion expected-values, where the DECODED characters are wanted (a real
 * newline, a real backslash, `café`).
 *
 * Ported for ADR-023 Stage 3 (#501).
 */
class PartialOutputExtractorTests {

    private val extractor = PartialOutputExtractor()

    // MARK: - Empty / pre-JSON

    @Test
    fun emptyBufferYieldsEmptySnapshot() {
        assertEquals(PartialSnapshot.empty, extractor.extract(""))
    }

    @Test
    fun bufferWithoutOpenBraceYieldsEmpty() {
        assertEquals(PartialSnapshot.empty, extractor.extract("Sure! Here is the JSON:"))
    }

    @Test
    fun bufferStoppedBeforeOpeningQuoteYieldsEmpty() {
        // Colon not yet arrived — extractor must wait.
        assertNull(extractor.extract("""{"statement"""").primary)
        // Colon arrived but no opening quote — still wait.
        assertNull(extractor.extract("""{"statement":""").primary)
    }

    // MARK: - Primary key reveals

    @Test
    fun primaryRevealsAfterOpeningQuote() {
        // Opening quote present but no content — primary is empty string, not null.
        val snap = extractor.extract("""{"statement":"""")
        assertEquals("", snap.primary)
    }

    @Test
    fun primaryRevealsIncrementalText() {
        val snap = extractor.extract("""{"statement":"Let's coope""")
        assertEquals("Let's coope", snap.primary)
        assertNull(snap.thought)
    }

    @Test
    fun primaryRevealsCompleteValue() {
        val snap = extractor.extract("""{"statement":"Let's cooperate."}""")
        assertEquals("Let's cooperate.", snap.primary)
    }

    @Test
    fun allKnownPrimaryKeysResolve() {
        for (key in PartialOutputExtractor.primaryKeys) {
            val snap = extractor.extract("""{"$key":"val"}""")
            assertEquals("val", snap.primary, "primary key $key failed to resolve")
        }
    }

    @Test
    fun reflectNoteStreamsAsPrimary() {
        // reflect's canonical primary field is `note`; a partial reflect JSON must
        // yield a live primary snapshot as it types in.
        val partial = extractor.extract("""{"note":"I should keep quiet abo""")
        assertEquals("I should keep quiet abo", partial.primary)
        val complete = extractor.extract("""{"note":"I should keep quiet."}""")
        assertEquals("I should keep quiet.", complete.primary)
    }

    // MARK: - Thought

    @Test
    fun thoughtResolvesAfterPrimary() {
        val snap = extractor.extract("""{"statement":"hi","inner_thought":"secret"}""")
        assertEquals("hi", snap.primary)
        assertEquals("secret", snap.thought)
    }

    @Test
    fun thoughtIsNilIfNotYetOpened() {
        val snap = extractor.extract("""{"statement":"hi"""")
        assertEquals("hi", snap.primary)
        assertNull(snap.thought)
    }

    // MARK: - Phase-specific thought key (#609)

    // The vote phase's private-thought field is `reason`, not `inner_thought`.
    // The caller passes the schema-derived thought key so the live streaming
    // THINKING section surfaces the vote reason as it types in (parity with
    // speak's `inner_thought`).
    @Test
    fun reasonThoughtKeyExtractsVoteReason() {
        val snap = extractor.extract(
            """{"vote":"Dave","reason":"散歩に行きたい"}""",
            thoughtKey = "reason",
        )
        assertEquals("Dave", snap.primary)
        assertEquals("散歩に行きたい", snap.thought)
    }

    @Test
    fun reasonThoughtStreamsIncrementally() {
        val snap = extractor.extract(
            """{"vote":"Dave","reason":"散""",
            thoughtKey = "reason",
        )
        assertEquals("Dave", snap.primary)
        assertEquals("散", snap.thought)
    }

    // Default thought key stays `inner_thought` (back-compat): a vote buffer read
    // with the default key surfaces no thought.
    @Test
    fun defaultThoughtKeyIgnoresReason() {
        val snap = extractor.extract("""{"vote":"Dave","reason":"散歩"}""")
        assertEquals("Dave", snap.primary)
        assertNull(snap.thought)
    }

    // MARK: - Escape handling

    @Test
    fun escapedQuoteInPrimary() {
        val snap = extractor.extract("""{"statement":"She said \"hi\""}""")
        assertEquals("She said \"hi\"", snap.primary)
    }

    @Test
    fun escapedBackslashInPrimary() {
        val snap = extractor.extract("""{"statement":"a\\b"}""")
        assertEquals("a\\b", snap.primary)
    }

    @Test
    fun escapedNewlineInPrimary() {
        val snap = extractor.extract("""{"statement":"line1\nline2"}""")
        assertEquals("line1\nline2", snap.primary)
    }

    @Test
    fun incompleteEscapeAtEndHoldsBack() {
        // Buffer ends with a lone backslash — the next char might be `"` (end of
        // string) or `\\` (literal backslash). Must not emit the `\` yet.
        val snap = extractor.extract("""{"statement":"x\""")
        assertEquals("x", snap.primary)
    }

    @Test
    fun incompleteUnicodeEscapeHoldsBack() {
        // \uXXXX needs 4 hex digits — fewer is incomplete.
        val snap = extractor.extract("""{"statement":"a\u00""")
        assertEquals("a", snap.primary)
    }

    @Test
    fun completeUnicodeEscapeDecodes() {
        // é is é. The raw-string buffer holds the literal `é` escape
        // sequence (Kotlin raw strings do not process \u escapes) for the extractor
        // to decode; the expected value is the decoded character.
        val snap = extractor.extract("""{"statement":"caf\u00e9"}""")
        assertEquals("café", snap.primary)
    }

    @Test
    fun signedUnicodeEscapeHoldsBack() {
        // Kotlin's toIntOrNull(16) accepts a leading '-'; Swift's UInt32(_:radix:)
        // rejects it. The hex-charset guard makes both hold back the incomplete
        // escape.
        val snap = extractor.extract("""{"statement":"a\u-0ff""")
        assertEquals("a", snap.primary)
    }

    @Test
    fun surrogateUnicodeEscapeHoldsBack() {
        // A lone surrogate code point (0xD800..0xDFFF) has no scalar value: Swift's
        // Unicode.Scalar(code) returns nil there, so the Swift original holds back.
        // GUARD 2 reproduces that — without it, code.toChar() would append a lone
        // surrogate Char and break parity. (Not in the Swift suite; added to pin the
        // ported guard under ADR-023 §12 condition-4 perturbation.)
        val snap = extractor.extract("""{"statement":"a\uD800"}""")
        assertEquals("a", snap.primary)
    }

    // MARK: - Thinking-tag handling

    @Test
    fun unclosedChannelThinkingTagHidesEverything() {
        // While we're inside a thinking tag, no extraction should fire even if `{`
        // appears later in the tag's content.
        val snap = extractor.extract("""<|channel>thought\nI'll say {"statement":"hi"}""")
        assertEquals(PartialSnapshot.empty, snap)
    }

    @Test
    fun closedChannelThinkingTagIsStripped() {
        val buffer = """
            <|channel>thought
            reasoning here
            <channel|>{"statement":"visible"}
        """.trimIndent()
        assertEquals("visible", extractor.extract(buffer).primary)
    }

    @Test
    fun unclosedThinkTagHidesEverything() {
        val snap = extractor.extract("""<think>reasoning {"statement":"hi""")
        assertEquals(PartialSnapshot.empty, snap)
    }

    // MARK: - Leading garbage

    @Test
    fun leadingGarbageSkippedUntilBrace() {
        val snap = extractor.extract("Sure! Here's my response:\n\n{\"statement\":\"OK\"}")
        assertEquals("OK", snap.primary)
    }
}
